File: sdks/python/src/aitelier_client/webhooks.py
"""Helpers for receiving aitelier webhooks.

Aitelier authenticates every webhook delivery with a pre-shared bearer token:

    Authorization: Bearer <webhook_secret>

…when `[service] webhook_secret` is set on the server. Receivers verify by
comparing the presented token to their own copy of the secret in constant time.

The function below is the entire contract on the receiver side. We ship it so
consumers don't reinvent it (and don't accidentally use `==` instead of
`hmac.compare_digest`, which leaks timing information).
"""

from __future__ import annotations

import hmac

_BEARER_PREFIX = "Bearer "


def verify_webhook_bearer(authorization_header: str | None, secret: str) -> bool:
    """True iff `authorization_header` is exactly `Bearer <secret>`.

    `authorization_header` is the raw `Authorization` header value as received
    over the wire (e.g. `"Bearer s3cr3t"`). Pass `None` if the header is absent —
    the function returns False rather than raising, so receivers can branch on
    the return.

    Uses `hmac.compare_digest` for a timing-safe comparison: a wall-clock
    attacker can't reconstruct the secret byte-by-byte by measuring response
    time. Bearer (not an HMAC body signature) is aitelier's delivery-auth
    mechanism — see `docs/INTEGRATION.md` and the server's `webhook_worker`.
    """
    if not authorization_header or not authorization_header.startswith(_BEARER_PREFIX):
        return False
    token = authorization_header[len(_BEARER_PREFIX):]
    return hmac.compare_digest(token, secret)

File: sdks/typescript/src/webhooks.ts
/**
 * Helpers for receiving aitelier webhooks.
 *
 * Aitelier authenticates every webhook delivery with a pre-shared bearer
 * token:
 *
 *     Authorization: Bearer <webhook_secret>
 *
 * …when `[service] webhook_secret` is set on the server. Receivers verify by
 * comparing the presented token to their own copy of the secret in constant
 * time.
 *
 * We ship this so consumers don't reinvent it (and don't accidentally use
 * `===` instead of `crypto.timingSafeEqual`, which leaks timing information).
 */
import { timingSafeEqual } from "node:crypto";

const BEARER_PREFIX = "Bearer ";

/**
 * True iff `authorizationHeader` is exactly `Bearer <secret>`.
 *
 * `authorizationHeader` is the raw `Authorization` header value as received
 * (e.g. `"Bearer s3cr3t"`). Pass `null`/`undefined` if the header is absent —
 * the function returns `false` rather than throwing.
 *
 * Uses `crypto.timingSafeEqual` for constant-time comparison so a wall-clock
 * attacker can't reconstruct the secret byte-by-byte by measuring response
 * time. Bearer (not an HMAC body signature) is aitelier's delivery-auth
 * mechanism — see `docs/INTEGRATION.md`.
 */
export function verifyWebhookBearer(
  authorizationHeader: string | null | undefined,
  secret: string,
): boolean {
  if (!authorizationHeader || !authorizationHeader.startsWith(BEARER_PREFIX)) {
    return false;
  }
  const token = Buffer.from(authorizationHeader.slice(BEARER_PREFIX.length), "utf8");
  const expected = Buffer.from(secret, "utf8");
  // timingSafeEqual requires equal-length Buffers. A length mismatch is
  // already a non-match; short-circuit to avoid a thrown exception (which
  // would itself be a side channel).
  if (token.length !== expected.length) {
    return false;
  }
  return timingSafeEqual(token, expected);
}
